presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayJoin.java

```diff
@@ -0,0 +1,216 @@
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type type = types.get("T");
+        TypeSignature arrayType = parameterizedTypeName(StandardTypes.ARRAY, type.getTypeSignature());
+        Signature signature = new Signature(FUNCTION_NAME, VARCHAR_TYPE_SIGNATURE, arrayType, VARCHAR_TYPE_SIGNATURE);
+        MethodHandle methodHandle = methodHandle(ArrayJoin.class, "arrayJoin", FunctionInfo.class, Type.class, ConnectorSession.class, Slice.class, Slice.class);
+        return specializeArrayJoin(types, functionRegistry, ImmutableList.of(false, false), signature, methodHandle);
+    }
+
+    private static FunctionInfo specializeArrayJoin(Map<String, Type> types, FunctionRegistry functionRegistry, List<Boolean> nullableArguments, Signature signature, MethodHandle methodHandle)
+    {
+        Type type = types.get("T");
+        FunctionInfo castFunction = functionRegistry.getExactFunction(internalOperator(CAST.name(), VARCHAR_TYPE_SIGNATURE, ImmutableList.of(type.getTypeSignature())));
+
+        if (!(type instanceof UnknownType)) {
+            checkCondition(castFunction != null, INVALID_FUNCTION_ARGUMENT, "Input type %s not supported", type);
+        }
+
+        return new FunctionInfo(signature, DESCRIPTION, false, methodHandle.bindTo(castFunction).bindTo(type), true, false, nullableArguments);
+    }
+
+    public static Slice arrayJoin(FunctionInfo castFunction, Type elementType, ConnectorSession session, Slice array, Slice delimiter, Slice nullReplacement) throws Throwable
+    {
+        return doArrayJoin(castFunction, elementType, session, array, delimiter, nullReplacement);
+    }
+
+    public static Slice arrayJoin(FunctionInfo castFunction, Type elementType, ConnectorSession session, Slice array, Slice delimiter) throws Throwable
+    {
+        return doArrayJoin(castFunction, elementType, session, array, delimiter, null);
+    }
+
+    private static Slice doArrayJoin(FunctionInfo castFunction, Type elementType, ConnectorSession session, Slice array, Slice delimiter, Slice nullReplacement)
+            throws Throwable
+    {
+        Block arrayBlock = readStructuralBlock(array);
+        int numElements = arrayBlock.getPositionCount();
+        DynamicSliceOutput sliceOutput = new DynamicSliceOutput(array.length());
+        Class<?> javaType = elementType.getJavaType();
+
+        Class<?>[] parameters = null;
+        MethodHandle castFunctionHandle = null;
+        if (castFunction != null) { //can be null for the unknown type
+            parameters = castFunction.getMethodHandle().type().parameterArray();
+            castFunctionHandle = castFunction.getMethodHandle();
+        }
+
+        for (int i = 0; i < numElements; i++) {
+            if (arrayBlock.isNull(i)) {
+                if (nullReplacement != null) {
+                    sliceOutput.appendBytes(nullReplacement);
+                }
```

Comment:
Instead of having `array_join(x, delimiter)` replace NULLs with the empty string, lets have it skip NULLs. So add an else branch that `continue`s. That is `array_join(ARRAY[1, NULL, 2], ',')` should be "1,2" instead of "1,,2"